data_model/src/executor.rs

```diff
@@ -47,10 +52,87 @@ mod model {
         pub wasm: WasmSmartContract,
     }
 
+    /// Executor data model.
+    ///
+    /// Defined from within the executor, it describes certain structures the executor
+    /// works with.
+    ///
+    /// Executor can define:
+    ///
+    /// - Permission tokens (see [`crate::permission::PermissionToken`])
+    /// - Configuration parameters (see [`crate::parameter::Parameter`])
+    #[derive(
+        Default,
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Constructor,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+        Display,
+    )]
+    #[ffi_type]
+    #[display(fmt = "{self:?}")]
+    pub struct ExecutorDataModel {
+        /// Data model JSON schema, typically produced by [`IntoSchema`].
+        pub schema: JsonString,
+        /// Permission tokens supported by the executor.
+        ///
+        /// These IDs refer to the types in the schema.
+        pub permission_token_ids: BTreeSet<PermissionTokenId>,
+        /// Configuration parameters supported by the executor.
+        ///
+        /// These IDs refer to the types in the schema.
+        pub parameter_ids: BTreeSet<ParameterId>,
+    }
+
     // TODO: Client doesn't need structures defined inside this macro. When dynamic linking is
     // implemented use: #[cfg(any(feature = "transparent_api", feature = "ffi_import"))]
 }
 
+impl ExecutorDataModel {
+    /// Getter
+    pub fn schema(&self) -> &JsonString {
+        &self.schema
+    }
+
+    /// Getter
+    pub fn permission_token_ids(&self) -> &BTreeSet<PermissionTokenId> {
+        &self.permission_token_ids
+    }
+
+    /// Getter
+    pub fn parameter_ids(&self) -> &BTreeSet<ParameterId> {
+        &self.parameter_ids
+    }
+}
+
+/// Defines an object that is part of the [`ExecutorDataModel`].
+///
+/// An object consists of id and payload.
+/// ID is a unique type identifier in the executor's data model.
+/// Payload is JSON-serialized type structure, according to the type schema.
+// TODO: support both JSON and SCALE for payload.
+pub trait ExecutorDataModelObject: Sized {
```

Comment:
why don't you put this trait into `smart_contract`?

Reply:
Indeed.